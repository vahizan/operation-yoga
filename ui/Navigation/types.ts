import { ReactNode } from "react";

export type NavItemType = {
  isRoot?: boolean;
  title: string;
  link: string;
  icon?: string | SVGElement | ReactNode;
  imageAlt?: string;
};

export type NavMenuType = {
  root: NavItemType;
  children?: NavMenuType[] | undefined;
  key?: string | number;
};
