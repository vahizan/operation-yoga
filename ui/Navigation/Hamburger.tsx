"use client";

import styles from "./Navigation.module.scss";
import {
  Dispatch,
  FocusEventHandler,
  MouseEventHandler,
  SetStateAction,
} from "react";

interface Props {
  setNavigationOpen: Dispatch<SetStateAction<boolean>>;
  isNavigationOpen: boolean;
}

export default function Hamburger({
  setNavigationOpen,
  isNavigationOpen,
}: Props) {
  const clickHandler: MouseEventHandler<HTMLInputElement> = (event) => {
    setNavigationOpen(!isNavigationOpen);
  };

  const onBlurHandler: FocusEventHandler<HTMLInputElement> = () => {
    setNavigationOpen(false);
  };

  return (
    <>
      <input
        onClick={clickHandler}
        onBlur={onBlurHandler}
        type="button"
        id="hamburger-button"
        data-test-id="hamburger-button"
        className={`${isNavigationOpen && styles.buttonClicked} ${
          styles.hamburgerButton
        } `}
      />
      <label htmlFor="hamburger-button">
        <div className={`${styles.hamburger}`}>
          <span className={`${styles.bar} ${styles.bar1}`}></span>
          <span className={`${styles.bar} ${styles.bar2}`}></span>
          <span className={`${styles.bar} ${styles.bar3}`}></span>
          <span className={`${styles.bar} ${styles.bar4}`}></span>
          <span className={`${styles.bar} ${styles.bar5}`}></span>
        </div>
      </label>
    </>
  );
}
