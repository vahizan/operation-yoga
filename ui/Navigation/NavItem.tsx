import styles from "./Navigation.module.scss";
import { ReactNode } from "react";
import { NavItemType } from "../../types/NavigationTypes";
import Link from "next/link";

const renderIcon = (
  icon: string | SVGElement | ReactNode,
  imageAlt: string
): ReactNode => {
  if (typeof icon === "string") {
    return <img src={icon} alt={imageAlt} />;
  } else {
    return <>{icon}</>;
  }
};
const renderItem = (
  title: string,
  link: string,
  icon: string | SVGElement | ReactNode,
  imageAlt: string
) => {
  return (
    <>
      {icon && imageAlt && (
        <span className={styles.navItem__icon}>
          {renderIcon(icon, imageAlt)}
        </span>
      )}
      <Link className={styles.navItem__button} href={link}>
        {title}
      </Link>
    </>
  );
};
export default function NavItem({
  title,
  link,
  icon,
  imageAlt = "image not available",
  isRoot,
}: NavItemType) {
  return isRoot ? (
    <div key={title} className={`${styles.navItem} ${styles.navItem__root}`}>
      {renderItem(title, link, icon, imageAlt)}
    </div>
  ) : (
    <li key={title} className={styles.navItem}>
      {renderItem(title, link, icon, imageAlt)}
    </li>
  );
}
