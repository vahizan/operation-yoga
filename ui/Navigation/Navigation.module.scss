@import "../../styles/breakpoints.scss";
@import "../../styles/utils.scss";


@keyframes shotup {
  0% {
    transform: translateY(300%) ;
    opacity: 0;
  }
  90% {
    transform: translateY(-50px);
  }
  100% {
    transform: translateY(0%) ;
    opacity: 1;
  }
}

@keyframes slideOut {
  0% {
    width: 0;
    opacity: 0;
  }

  100% {
    width: 30%;
    opacity: 1;
  }
}

@keyframes slideIn {
  0% {
    width: 30%;
    // transform: scale(.3);
    opacity: 1;
  }

  100% {
    width: 0%;
    // transform:scale(1);
    opacity: 0;
  }
}


.navigationContainer {
  top: 0;
  right: 0;
  left: 0;
  z-index: 1030;
  width: 100%;
  display: flex;
  justify-content: center;
  position: sticky;
  height: 100px;
}

.navigation {
  height: 100%;
  display: flex;
  justify-content: center;
  flex-direction: row;

  ul {
    padding: 0;
    margin: 0;
    align-self: center;
  }
}

.logo{
  align-self: center;
  width: 110px;
  flex: 0 0;
}


.navMenu {
  @include desktop-up {
    width:150px;
  }

  text-align: center;
  padding-top: 20px;
  padding-bottom: 20px;
  display: flex;
  flex-direction: column;

  &__root {
    &__children {
      padding-left: 20px;
      padding-right: 20px;
      opacity: 0;
    }

    &:hover &__children {
      background: white;
      opacity: 1;
    }
  }

}

.navItem {
  display: flex;
  justify-content:center ;
  text-align: center;
  width: 100%;
  padding: 20px;
  list-style-type: none;

  &__button {
    padding: 15px 25px;
    width: 100%;

  }
  &__root {
    a {
      @include autoFontSize(25);
    }
  }

  button {
    padding: inherit;
    @include autoFontSize(17);
    font-weight: bolder;
    width: 100%;
    color: initial;
    background-color: initial;
    border: none;
    cursor: pointer;

    a {
      display: block;
    }
  }

  @include desktop-up {
    &:hover {
      display: block;
      background-image: $glow-bg;
      transition: opacity .3s ease;
      z-index: 2;
      border-radius: 10px;
    }
  }

}


.verticalNavMenu {
  text-align: center;
  padding-top: 10px;
  padding-bottom: 10px;
  display: flex;
  flex-direction: column;

  &__root {
    @include autoFontSize(24);
    &__children {
      @include autoFontSize(18);
      padding-left: 20px;
      padding-right: 20px;
    }
  }

}

.leftDrawer__hamburger{
  &:hover {
    background-image: $glow-bg;
  }

  display: flex;
  align-items: center;
  justify-content: center;
  background-color: white;
  border-radius: 100px;
  position: fixed;
  z-index: 100;
  height: 60px;
  width: 60px;
  left: 20px;
  top: 20px;
}

.leftDrawer__content {
  overflow-y: auto;
  padding-top: 50px;
  @include shadow();
  background-color: white;
  position: fixed;
  left: 0;
  height: 100%;
  width: 0;
  z-index: 10;
  opacity: 0;
  transition: all 600ms cubic-bezier(.62,.04,.3,1.56);
  transition-delay: 100ms;
}

.leftDrawer__content__show {
  @include mobile-up {
    width: 100%;
  }

  @include desktop-up {
    width:30%
  }
  opacity: 1;
}

.leftDrawer__content__hide {
  opacity: 0;
}


.verticalNavItem {
  width: 100%;
  padding: 10px;
  list-style-type: none;

  @include desktop-up {
    &:hover {
      display: block;
      background-image: $glow-bg;
      transition: opacity .3s ease;
      z-index: 2;
      border-radius: 10px;
    }
  }

}

.hidden {
  position: absolute;
  overflow: hidden;
  clip: rect(0 0 0 0);
  height: 1px; width: 1px;
  margin: -1px; padding: 0; border: 0;
}

.hamburger {
  top: -5px;
  width: 30px;
  height: 30px;
  position: relative;
}

.hamburger .bar{
  padding: 0;
  top: 20px;
  width: 30px;
  height: 4px;
  background-color: black;
  display: block;
  border-radius: 4px;
  transition: all 0.4s ease-in-out;
  position: absolute;
}

.hamburger .bar {
  transition: all 0.4s ease-in-out, transform 0.4s ease-in-out 0.4s;
}


.bar1 {
  top: 0;
}

.bar2,
.bar3 {
  top: 13.5px;
}

.bar3 {
  right: 0;
}

.hamburger .bar4 {
  top: 10px;
}

.hamburger .bar5 {
  top: 30px;
}

.hamburger .bar {
  transition: all 0.4s ease-in-out, transform 0.4s ease-in-out 0.4s;
}

.hamburger .bar2 {
  width: 1px;
  transform: rotate(90deg);
  left: 13.5px;
}

.hamburger .bar3 {
  width: 1px;
  left: 13.5px;

}
.hamburgerButton {
  background: transparent;
  border: none;
  position: absolute;
  width: 100%;
  height: 100%;
  z-index: 10;
}

.buttonClicked + label > .hamburger > .bar1{
  top: 13.5px;
  background-color: transparent;
}

.buttonClicked + label > .hamburger > .bar2{
  left: 0px;
  width: 30px;
  transform: rotate(45deg);
}

.buttonClicked + label > .hamburger > .bar3{
  left: 0;
  width: 30px;
  transform: rotate(-45deg);
}

.buttonClicked + label > .hamburger > .bar4{
  background-color: transparent;
}

.buttonClicked + label > .hamburger > .bar5{
  bottom: 13.5px;
  background-color: transparent;
}