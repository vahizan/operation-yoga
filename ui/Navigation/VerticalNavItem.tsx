import styles from "./Navigation.module.scss";
import { ReactNode } from "react";
import { NavItemType } from "../../types/NavigationTypes";

const renderIcon = (
  icon: string | SVGElement | ReactNode,
  imageAlt: string
): ReactNode => {
  if (typeof icon === "string") {
    return <img src={icon} alt={imageAlt} />;
  } else {
    return <>{icon}</>;
  }
};
const renderItem = (
  title: string,
  link: string,
  icon: string | SVGElement | ReactNode,
  imageAlt: string
) => {
  return (
    <>
      {icon && imageAlt && (
        <span className={styles.verticalNavItem__icon}>
          {renderIcon(icon, imageAlt)}
        </span>
      )}
      <a href={link} className={styles.verticalNavItem__link}>
        {title}
      </a>
    </>
  );
};
export default function VerticalNavItem({
  title,
  link,
  icon,
  imageAlt = "image is not available",
  isRoot,
}: NavItemType) {
  return isRoot ? (
    <div
      key={title}
      className={`${styles.verticalNavItem} ${styles.verticalNavItem__root}`}
    >
      {" "}
      {renderItem(title, link, icon, imageAlt)}
    </div>
  ) : (
    <li key={title} className={styles.verticalNavItem}>
      {renderItem(title, link, icon, imageAlt)}
    </li>
  );
}
