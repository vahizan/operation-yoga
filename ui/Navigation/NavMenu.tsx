import styles from "./Navigation.module.scss";
import {NavMenuType} from "./types";
import NavItem from "./NavItem";


export default function NavMenu({root, children}: NavMenuType ) {
    if(!children && !root){
        return;
    }
    if(!children && root){
        return <NavItem title={root.title} link={root.link} icon={root?.icon} imageAlt={root?.imageAlt}/>
    }
    return (
        <div className={styles.navMenu}>
            <div className={styles.navMenu__root}>
                <NavItem key={`${root.title}_root`} isRoot title={root.title} link={root.link} icon={root?.icon} imageAlt={root?.imageAlt}/>
                <ul className={styles.navMenu__root__children}>
                    {children.map((child) => {
                        return NavMenu(child);
                    })}
                </ul>
            </div>
        </div>
    );
}