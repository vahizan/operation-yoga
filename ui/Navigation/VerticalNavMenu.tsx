import styles from "./Navigation.module.scss";
import { NavMenuType } from "./types";
import NavItem from "./NavItem";

export default function VerticalNavMenu({ root, children, key }: NavMenuType) {
  if (!children && !root) {
    return null;
  }
  if (!children && root) {
    return (
      <NavItem
        key={key}
        title={root.title}
        link={root.link}
        icon={root?.icon}
        imageAlt={root?.imageAlt}
      />
    );
  }
  return (
    <div className={styles.verticalNavMenu}>
      <div className={styles.verticalNavMenu__root}>
        <NavItem
          key={`${root.title}_root`}
          isRoot
          title={root.title}
          link={root.link}
          icon={root?.icon}
          imageAlt={root?.imageAlt}
        />
        <ul className={styles.verticalNavMenu__root__children}>
          {children &&
            children.map((child) => {
              return VerticalNavMenu(child);
            })}
        </ul>
      </div>
    </div>
  );
}
