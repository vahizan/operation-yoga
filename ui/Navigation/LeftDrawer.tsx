"use client";

import styles from "./Navigation.module.scss";
import Hamburger from "./Hamburger";
import { NavMenuType } from "./types";
import { NAVIGATION_MENU_VALUES } from "../../constants/Navigation.constants";
import VerticalNavMenu from "./VerticalNavMenu";
import { useState } from "react";

const renderNavItems = (navMenuItems: NavMenuType[]) => {
  if (!navMenuItems) {
    return;
  }
  return navMenuItems.map((navMenu, i) => {
    return (
      <VerticalNavMenu
        root={navMenu.root}
        children={navMenu.children}
        key={`${navMenu.root.title}_${i}`}
      />
    );
  });
};

export default function LeftDrawer() {
  const [isChecked, setChecked] = useState(false);
  return (
    <div className={styles.leftDrawer}>
      <div className={styles.leftDrawer__hamburger}>
        <Hamburger updateValue={setChecked} />
      </div>

      <div
        className={`${styles.leftDrawer__content}
          ${
            isChecked
              ? styles.leftDrawer__content__show
              : styles.leftDrawer__content__hide
          }
            `}
      >
        {renderNavItems(NAVIGATION_MENU_VALUES)}
      </div>
    </div>
  );
}
