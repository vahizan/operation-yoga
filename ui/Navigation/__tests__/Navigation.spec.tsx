import { screen, render } from '@testing-library/react';
import Navigation from "../../Navigation";

import {NAVIGATION_MENU_VALUES} from "../../../constants/Navigation.constants";

describe("Navigation", () => {
    const {container} = render(<Navigation navItems={NAVIGATION_MENU_VALUES} />);
    describe("generate nav links", () => {
        it('should create nav links from config', () => {
            expect(container).toMatchSnapshot();
        });
    });
});