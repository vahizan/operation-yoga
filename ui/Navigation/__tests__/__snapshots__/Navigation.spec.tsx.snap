// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation generate nav links should create nav links from config 1`] = `
<div>
  <div
    class="navigationContainer"
  >
    <nav
      class="navigation"
    >
      <div
        class="navMenu"
      >
        <div
          class="navMenu__root"
        >
          <div
            class="navItem navItem__root"
          >
            <a
              class="navItem__button"
              href="/"
            >
              Home
            </a>
          </div>
          <ul
            class="navMenu__root__children"
          />
        </div>
      </div>
      <li
        class="navItem"
      >
        <a
          class="navItem__button"
          href="/location"
        >
          Location
        </a>
      </li>
      <li
        class="navItem"
      >
        <a
          class="navItem__button"
          href="/about"
        >
          About Us
        </a>
      </li>
      <li
        class="navItem"
      >
        <a
          class="navItem__button"
          href="/our-mission"
        >
          Our Mission
        </a>
      </li>
      <li
        class="navItem"
      >
        <a
          class="navItem__button"
          href="/teachers"
        >
          Our Teachers
        </a>
      </li>
      <li
        class="navItem"
      >
        <a
          class="navItem__button"
          href="/contact-us"
        >
          Contact Us
        </a>
      </li>
      <li
        class="navItem"
      >
        <a
          class="navItem__button"
          href="/quick-booking"
        >
          Quick Booking
        </a>
      </li>
      <div
        class="navMenu"
      >
        <div
          class="navMenu__root"
        >
          <div
            class="navItem navItem__root"
          >
            <a
              class="navItem__button"
              href="/veda-yogshala"
            >
              Veda Yogshala
            </a>
          </div>
          <ul
            class="navMenu__root__children"
          >
            <li
              class="navItem"
            >
              <a
                class="navItem__button"
                href="/veda-yogshala/yoga-classes"
              >
                Yoga Classes
              </a>
            </li>
            <li
              class="navItem"
            >
              <a
                class="navItem__button"
                href="/veda-yogshala/retreats"
              >
                Retreats
              </a>
            </li>
            <li
              class="navItem"
            >
              <a
                class="navItem__button"
                href="/veda-yogshala/teacher-training-courses"
              >
                Teachers' Training Courses
              </a>
            </li>
            <li
              class="navItem"
            >
              <a
                class="navItem__button"
                href="/veda-yogshala/language-courses"
              >
                Language Courses
              </a>
            </li>
            <li
              class="navItem"
            >
              <a
                class="navItem__button"
                href="/veda-yogshala/indian-cooking-classes"
              >
                Indian Cooking Classes
              </a>
            </li>
            <li
              class="navItem"
            >
              <a
                class="navItem__button"
                href="/veda-yogshala/pranayama-meditation"
              >
                Pranayama & Meditation
              </a>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="navMenu"
      >
        <div
          class="navMenu__root"
        >
          <div
            class="navItem navItem__root"
          >
            <a
              class="navItem__button"
              href="/veda-spa"
            >
              Veda Spa
            </a>
          </div>
          <ul
            class="navMenu__root__children"
          >
            <li
              class="navItem"
            >
              <a
                class="navItem__button"
                href="/veda-spa/spa-treatments"
              >
                Spa Treatments
              </a>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="navMenu"
      >
        <div
          class="navMenu__root"
        >
          <div
            class="navItem navItem__root"
          >
            <a
              class="navItem__button"
              href="/blog"
            >
              Blog
            </a>
          </div>
          <ul
            class="navMenu__root__children"
          >
            <li
              class="navItem"
            >
              <a
                class="navItem__button"
                href="/blog/what-is-hatha-yoga"
              >
                What is Hatha Yoga
              </a>
            </li>
            <li
              class="navItem"
            >
              <a
                class="navItem__button"
                href="/blog/more-about-ashtanga"
              >
                More about Ashtanga
              </a>
            </li>
            <li
              class="navItem"
            >
              <a
                class="navItem__button"
                href="/blog/mountain-pose-vs-downward-dog"
              >
                Mountain Pose vs Downward Dog
              </a>
            </li>
          </ul>
        </div>
      </div>
      <div
        class="navMenu"
      >
        <div
          class="navMenu__root"
        >
          <div
            class="navItem navItem__root"
          >
            <a
              class="navItem__button"
              href="/gallery"
            >
              Gallery
            </a>
          </div>
          <ul
            class="navMenu__root__children"
          >
            <li
              class="navItem"
            >
              <a
                class="navItem__button"
                href="/gallery/photo"
              >
                Photo
              </a>
            </li>
            <li
              class="navItem"
            >
              <a
                class="navItem__button"
                href="/gallery/video"
              >
                Video
              </a>
            </li>
          </ul>
        </div>
      </div>
    </nav>
  </div>
</div>
`;
